jax/lax/lax.py

```diff
@@ -3685,20 +3685,90 @@ def _scatter_mul_jvp_rhs(g, x, i, y, *, dimension_numbers, **kw):
 batching.primitive_batchers[scatter_mul_p] = (
   partial(_scatter_batching_rule, scatter_mul))
 
-# TODO(jlebar): Add derivatives.
+def _scatter_extremal_jvp(scatter_op, primals, tangents, update_jaxpr,
+                          update_consts, dimension_numbers):
+  operand, scatter_indices, updates = primals
+  g_operand, g_scatter_indices, g_updates = tangents
+
+  scatter_dnums = dimension_numbers
+  updates_shape = updates.shape
+
+  val_out = scatter_op.bind(
+      operand, scatter_indices, updates, update_jaxpr=update_jaxpr,
+      update_consts=update_consts, dimension_numbers=scatter_dnums)
+
+  if g_operand is ad_util.zero and g_updates is ad_util.zero:
+    tangent_out = ad_util.zero
+  else:
+    g_operand = ad.instantiate_zeros(operand, g_operand)
+    g_updates = ad.instantiate_zeros(updates, g_updates)
+
+    # gather_dnums and slice_sizes define the gather op that is the inverse of
+    # the scatter op specified by scatter_dnums
+    gather_dnums = GatherDimensionNumbers(
+        offset_dims=scatter_dnums.update_window_dims,
+        collapsed_slice_dims=scatter_dnums.inserted_window_dims,
+        start_index_map=scatter_dnums.scatter_dims_to_operand_dims)
+
+    slice_sizes = []
+    pos = 0
+    for i in range(len(operand.shape)):
+      if i in scatter_dnums.inserted_window_dims:
+        slice_sizes.append(1)
+      else:
+        slice_sizes.append(updates_shape[scatter_dnums.update_window_dims[pos]])
+        pos += 1
+
+    # For consistency with other max operations, if there are two or more values
+    # in updates that are contending to replace the same index location, the
+    # resulting tangent at that location will be the average of the associated
+    # tangents for the values in updates.
+    target_vals = gather(
+        val_out, scatter_indices, gather_dnums, onp.array(slice_sizes))
+
+    actual_updates = select(updates == target_vals, updates, _zeros(updates))
+
+    summed_updates = gather(
+        scatter_add(_zeros(operand),
+                    scatter_indices,
+                    actual_updates,
+                    scatter_dnums),
+        scatter_indices,
+        gather_dnums,
+        onp.array(slice_sizes))
+
+    coef = select((actual_updates == 0) | (summed_updates == 0),
+                  _ones(updates),
+                  actual_updates / summed_updates)
+
+    # This can be simplified once scatter has transpose implemented
+    target_tangents = gather(
+        g_operand, scatter_indices, gather_dnums, onp.array(slice_sizes))
+
+    tangent_updates = select(updates == target_vals,
```

Comment:
TODO: This actually won't correctly account for the case where an updates value is scattered to a location with the same value.

Reply:
Does this TODO mean the PR isn't ready for review, or would you like to get it in without this support first?

Reply:
Yes - hold the review, I'll fix this on Monday as it only occurred to me after I'd submitted. It won't fundamentally change the CL though so any general comments are welcome.